### PR TITLE
feat: Extension inference for conditional nodes

### PR DESCRIPTION
**Pull request by croyzor**
This builds on from #463 (merged) and #461.
Generate constraints saying that `Case` nodes' interfaces should match those of their `Conditional` and add a test

### CODE REVIEW

File: src/hugr.rs [hunk 1]
```diff
@@ -447,4 +467,55 @@ mod test {
         let hugr = simple_dfg_hugr();
         assert_matches!(hugr.get_io(hugr.root()), Some(_));
     }
+
+    #[test]
+    fn extension_instantiation() -> Result<(), Box<dyn Error>> {
+        const BIT: Type = crate::extension::prelude::USIZE_T;
```
Comment (ss2165) [inline src/hugr.rs:473]:
```suggestion
        const NAT: Type = crate::extension::prelude::USIZE_T;
```

to avoid confusion

File: src/hugr.rs [hunk 2]
```diff
@@ -447,4 +467,55 @@ mod test {
...
+
+        assert_eq!(
+            hugr.op_types
+                .get(lift.index)
+                .signature()
+                .unwrap()
+                .input_extensions,
+            ExtensionSet::new()
+        );
+        assert_eq!(
```
Comment (ss2165) [inline src/hugr.rs:511]: this test doesn't look related to the rest of the PR, worth adding a drive-by note

**Review by ss2165 — APPROVED** [review]
👍

File: src/hugr.rs (hunk 2 above)
Comment (croyzor) [reply src/hugr.rs:511]: I think this is a result of being based on #461. Now that that's merged I'll merge from `main` and expect this part of the diff to go away

**Comment by croyzor** [comment]
I've refactored the test to avoid using builder methods, since we want the parent `Conditional` to have the empty resource set and in the near future the builder will default to making it open. Unfortunately, this means losing access to the `CaseBuilder` too

**Review by ss2165 — APPROVED** [review]
